design-decisions/2020-02-06-One-Path-Deterministic.md

```diff
@@ -0,0 +1,190 @@
+endmodule
+```
+
+the one-path reachability claim 
+```
+<k> select => ?X </k> <state> SetItem(a) SetItem(b) SetItem(c) </state>
+    ensures ?X ==K b orBool ?X ==K c 
+```
+
+might not be provable because the prover selects to advance the execution
+using the `X=a` instance for the `select` rule.
+
+
+Options
+-------
+
+We could devise an algorithm to perform a more exhaustive exploration of the
+search-space.
+
+We will just sketch it below on an example.
+
+Consider the following specification
+```
+module TEST
+  imports BOOL
+  syntax S ::= "b" | "c" | "d" | "e"
+             | test(Bool)
+  
+  test(true) => b   // first rule
+  test(true) => c   // 2nd rule
+  test(false) => d  // 3rd rule
+  test(false) => e  // 4th rule
+endmodule
+```
+
+together with the claim `test(X) => c \/ e`.
+
+The idea would be to derive the goal in parallel with each of the rules,
+but to associate a remainder to each particular derivation,
+then we will require that the derived goal and the remainder need to be
+satisfied simultaneously, but that any of thus obtained
+derived goal - remainder pairs is sufficient to prove our original goal.
+
+Thus, we would replace `test(X) => c \/ e` by the derivation
+```
+(
+  b => c \/ e                     -- derivation using 1st rule
+  /\
+  test(X) /\ x <> true => c \/ e  -- remainder after applying 1st rule
+)
+\/
+(
+  c => c \/ e                     -- derivation using 2nd rule
+  /\
+  test(X) /\ x <> true => c \/ e  -- remainder after applying 2nd rule
+)
+\/
+(
+  d => c \/ e                     -- derivation using 3rd rule
+  /\
+  test(X) /\ x <> false => c \/ e -- remainder after applying 3rd rule
+)
+\/
+(
+  e => c \/ e                     -- derivation using 1st rule
+  /\
+  test(X) /\ x <> false => c \/ e -- remainder after applying  rule
+)
+```
+
+then we can choose any goal in the or-and tree and expand it, and continue to do
+so until all goals are stuck or one of the global disjuncts is completely proven.
+
+For example, assuming a leftmost, innermost strategy, we could continue
+by first checking `b => c \/ e` and declaring it stuck, meaning that we
+can skip its remainder altogether, then moving to `c => c \/ e` which
+holds without further expansion, so it can be eliminated, leading us to
+its remainder `test(X) /\ x <> true => c \/ e`.
+
+Following the algorithm, this can be expanded to
+```
+(
+  d => c \/ e
+  /\
+  test(X) /\ x <> true /\ x <> false => c \/ e
+)
+\/
+(
+  e => c \/ e
+  /\
+  test(X) /\ x <> true /\ x <> false => c \/ e
+)
+```
+
+Again, `d => c \/ e` is stuck, so we can skip to
+`e => c \/ e`, which holds immediately, so it can be eliminated.
+Its remainder, too, can be eliminated, as 
+` x <> true /\ x <> false` is equivalent to `⊥`.
+
+This concludes the proof, as we have found witnesses of reaching the
+destination for all instances of the input.
+
+### Note
+
+When deriving goals, care should be taken about the same rule unifing in
```

Comment:
I think in the current implementation of one path, it will consider all the unification instances instead of picking one of them. In other words, it consider unification results as `and` instead of `or`.

Reply:
I'm not 100% sure about that; An interesting anecdote that @ana-pantilie  can confirm is that in the `select` example above, attempting to prove that `select` can rewrite into `a` succeeds, while attempting to prove that it rewrites to either `b` or `c` fails...

Reply:
The following claim can be proven:
```  rule <k> select => ?X </k>
       <state> SetItem(a) SetItem(b) </state> 
    ensures ?X ==K a [one-path]
```
while if we switch `a` for `b` we get the following result:
```<generatedTop>
  <k>
    a
  </k>
  <state>
    SetItem ( a )
    SetItem ( b )
  </state>
</generatedTop>
```
@traiansf and I spent quite a while debugging this, and we also observed that in the case of `all-path` it behaves as expected, which led us to believe that our `one-path` implementation is the culprit.

Reply:
@traiansf This is a bug, right? Shouldn't one-path implementation consider all the unifiers?

Reply:
yet, it should , but it should make a disjunction between them.